### PR TITLE
grpcproxy/cache: acquire read lock on Get instead of write lock

### PR DESCRIPTION
**Pull request by heyitsanthony**
Otherwise can't scale past 1 CPU.

### CODE REVIEW

**Comment by heyitsanthony** [comment]
On further investigation, this particular fix isn't correct since the LRU needs exclusive access. Needs more thought.

**Comment by heyitsanthony** [comment]
Vendored a better LRU library. PTAL /cc @xiang90 @gyuho @fanminshi

**Comment by gyuho** [comment]
LGTM. Defer to @xiang90 

Q. how slow was it with the `Lock` vs. `RLock` in this patch?

File: proxy/grpcproxy/cache/store.go [hunk 1]
```diff
@@ -17,24 +17,28 @@ package cache
...
 
+const defaultHistoricTTL = time.Hour
+const defaultCurrentTTL = time.Minute
+
 type Cache interface {
 	Add(req *pb.RangeRequest, resp *pb.RangeResponse)
 	Get(req *pb.RangeRequest) (*pb.RangeResponse, error)
 	Compact(revision int64)
 	Invalidate(key []byte, endkey []byte)
-	Size() int
```
Comment (fanminshi) [inline proxy/grpcproxy/cache/store.go]: just wondering why is the Size() method removed?

File: proxy/grpcproxy/metrics.go [hunk 2]
```diff
@@ -29,12 +29,6 @@ var (
 		Name:      "events_coalescing_total",
 		Help:      "Total number of events coalescing",
 	})
-	cacheKeys = prometheus.NewGauge(prometheus.GaugeOpts{
```
Comment (fanminshi) [inline proxy/grpcproxy/metrics.go]: why is this removed as well?

File: proxy/grpcproxy/cache/store.go (hunk 1 above)
Comment (heyitsanthony) [inline proxy/grpcproxy/cache/store.go]: The concurrent LRU doesn't have a way to get the number of keys

File: proxy/grpcproxy/metrics.go (hunk 2 above)
Comment (heyitsanthony) [inline proxy/grpcproxy/metrics.go]: Since there's no way to get the number of keys, this metric can't be computed efficiently.

File: proxy/grpcproxy/cache/store.go [hunk 3]
```diff
@@ -73,7 +77,11 @@ func (c *cache) Add(req *pb.RangeRequest, resp *pb.RangeResponse) {
 	defer c.mu.Unlock()
 
 	if req.Revision > c.compactedRev {
-		c.lru.Add(key, resp)
+		if req.Revision == 0 {
+			c.lru.Set(key, resp, defaultCurrentTTL)
+		} else {
+			c.lru.Set(key, resp, defaultHistoricTTL)
+		}
```
Comment (fanminshi) [inline proxy/grpcproxy/cache/store.go]: could you explain the rationale behind using defaultCurrentTTL when req.Revision == 0 and defaultHistoricTTL when req.Revision != 0?

Comment (heyitsanthony) [inline proxy/grpcproxy/cache/store.go]: Historic keys are immutable so the TTL can be indefinite. Keys on the current revision are subject to change, so expiring on the TTL will refresh against the current server state.

Comment (fanminshi) [inline proxy/grpcproxy/cache/store.go]: is it possible the req.Revision from the req is the latest revision, eg kv store revision = 100 and req.Revision ==100? then Keys on that revision is the latest and are subject to change? if that's possible, does it means that we need to use defaultCurrentTTL then?

Comment (heyitsanthony) [inline proxy/grpcproxy/cache/store.go]: @fanminshi The cache is keyed by the request data. Get(rev=100) will always return rev=100; it is not subject to change (ignoring, rev=100 doesn't exist yet or rev=100 is compacted away, which is already checked). Get(rev=0) will not always return rev=100; it is subject to change.

Comment (fanminshi) [inline proxy/grpcproxy/cache/store.go]: @heyitsanthony gotcha, thanks for the clarification.

**Comment by heyitsanthony** [comment]
@gyuho tried to get a perf difference out of it, but nothing. Oh well.

**Comment by fanminshi** [comment]
LGTM

**Comment by gyuho** [comment]
@heyitsanthony as long as we can use more than 1 CPU :)

**Comment by heyitsanthony** [comment]
@gyuho yeah, it's just not contending on the lock yet. I imagine it'll be a problem if the proxy is running with a bunch of CPUs, but on the other hand it's stateless so it's always possible to spin up more processes if there is any contention on that lock.

Kind of luke-warm on this patch now that it's obvious the perf just isn't there, but it's got TTL's, so maybe still worthwhile switching out the lru implementation.

File: proxy/grpcproxy/cache/store.go [hunk 4]
```diff
@@ -73,7 +77,11 @@ func (c *cache) Add(req *pb.RangeRequest, resp *pb.RangeResponse) {
 	defer c.mu.Unlock()
 
 	if req.Revision > c.compactedRev {
-		c.lru.Add(key, resp)
+		if req.Revision == 0 {
```
Comment (xiang90) [inline proxy/grpcproxy/cache/store.go]: if req.rev ==0, we can actually add two entries. one for req.Revision = 0 and the other for req.Revision=resp.Revision.

**Comment by xiang90** [comment]
LGTM

File: proxy/grpcproxy/cache/store.go (hunk 4 above)
Comment (heyitsanthony) [inline proxy/grpcproxy/cache/store.go]: true; just preserving existing behavior for the time being
